test/distributed/fsdp/test_fsdp_state_dict.py

```diff
@@ -0,0 +1,127 @@
+# Owner(s): ["oncall: distributed"]
+
+import sys
+from typing import Any, Dict
+
+import torch
+from torch import distributed as dist
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.nn import Linear, Module
+from torch.nn.parallel import DistributedDataParallel
+from torch.optim import SGD
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import (
+    FSDPTest,
+    get_full_params,
+)
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+    TEST_WITH_DEV_DBG_ASAN,
+)
+
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+INNER_SHAPE = [4, 4]
+OUTER_SHAPE = [4, 5]
+
+
+class Model(Module):
```

Comment:
Wonder if we should test a more complicated set of models. A few ideas:

1. Nested FSDP module
2. FSDP module where shards need padding, i.e. total # of params is not divisible by the world size and we pad the shards. Seems like there's some specific logic there that we'd like to verify.
3. More complex module such as those with shared params, Transformer module, etc. These can be added as part of FSDP vs DDP parity test as I mentioned in another comment.

Reply:
https://github.com/pytorch/pytorch/issues/73048